internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go

```diff
@@ -0,0 +1,435 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"path/filepath"
+	"strings"
+
+	"github.com/AlekSi/pointer"
+	"github.com/talos-systems/os-runtime/pkg/controller"
+	"github.com/talos-systems/os-runtime/pkg/resource"
+	"github.com/talos-systems/os-runtime/pkg/state"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/talos-systems/talos/internal/app/machined/pkg/resources/config"
+	"github.com/talos-systems/talos/internal/app/machined/pkg/resources/k8s"
+	"github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1/machine"
+	"github.com/talos-systems/talos/pkg/machinery/constants"
+)
+
+// ControlPlaneStaticPodController manages k8s.StaticPod based on control plane configuration.
+type ControlPlaneStaticPodController struct {
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *ControlPlaneStaticPodController) Name() string {
+	return "k8s.ControlPlaneStaticPodController"
+}
+
+// ManagedResources implements controller.Controller interface.
+func (ctrl *ControlPlaneStaticPodController) ManagedResources() (resource.Namespace, resource.Type) {
+	return k8s.ControlPlaneNamespaceName, k8s.StaticPodType
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint: gocyclo
+func (ctrl *ControlPlaneStaticPodController) Run(ctx context.Context, r controller.Runtime, logger *log.Logger) error {
+	if err := r.UpdateDependencies([]controller.Dependency{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.K8sControlPlaneType,
+			Kind:      controller.DependencyWeak,
+		},
+		{
+			Namespace: k8s.ControlPlaneNamespaceName,
+			Type:      k8s.SecretsStatusType,
+			ID:        pointer.ToString(k8s.StaticPodSecretsStaticPodID),
+			Kind:      controller.DependencyWeak,
+		},
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineTypeType,
+			ID:        pointer.ToString(config.MachineTypeID),
+			Kind:      controller.DependencyWeak,
+		},
+	}); err != nil {
+		return fmt.Errorf("error setting up dependencies: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		machineTypeRes, err := r.Get(ctx, resource.NewMetadata(config.NamespaceName, config.MachineTypeType, config.MachineTypeID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting machine type: %w", err)
+		}
+
+		machineType := machineTypeRes.(*config.MachineType).MachineType()
+
+		if machineType != machine.TypeControlPlane && machineType != machine.TypeInit {
+			if err = ctrl.teardownAll(ctx, r); err != nil {
+				return fmt.Errorf("error destroying static pods: %w", err)
+			}
+		}
```

Comment:
Should this condition interrupt further loop execution?

Reply:
yep, it is missing `continue`, I believe I fixed in the recent push, but I need to recheck